pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+            during the TLS handshake. This may be advertised using either or
+            both of ALPN or NPN.
+
+            ``protocols`` should be a list of acceptable protocols in the form
+            of ``NextProtocol`` objects, such as ``[H2, HTTP1]``, ordered by
+            preference. The selection of the protocol will happen during the
+            handshake, and will use whatever protocol negotiation mechanisms
+            are available and supported by both peers.
+
+            If the TLS implementation doesn't support protocol negotiation,
+            this method will raise ``NotImplementedError``.
+            """
+
+        @abstractmethod
+        def set_sni_callback(self, callback: Optional[ServerNameCallback]) -> None:
+            """
+            Register a callback function that will be called after the TLS
+            Client Hello handshake message has been received by the TLS server
+            when the TLS client specifies a server name indication.
+
+            Only one callback can be set per ``Context``. If ``callback`` is
+            ``None`` then the callback is disabled. Calling this function a
+            subsequent time will disable the previously registered callback.
+
+            The ``callback`` function will be called with three arguments: the
+            first will be the ``TLSBufferObject`` for the connection; the
+            second will be a string that represents the server name that the
+            client is intending to communicate (or ``None`` if the TLS Client
+            Hello does not contain a server name); and the third argument will
+            be the original ``Context``. The server name argument will be the
+            IDNA *decoded* server name.
+
+            The ``callback`` must return ``None`` to allow negotiation to
+            continue. Other return values signal errors. Attempting to control
+            what error is signaled by the underlying TLS implementation is not
+            specified in this API, but is up to the concrete implementation to
+            handle.
+            """
+
+        @abstractmethod
+        def set_version_range(self, lower_bound=None: Optional[TLSVersion],
+                              upper_bound=None: Optional[TLSVersion]) -> None:
+            """
+            Set the minumum and maximum versions of TLS that should be allowed
+            on TLS connections made by this context.
+
+            If present, ``lower_bound`` will set the lowest acceptable TLS
+            version. If present, ``upper_bound`` will set the highest
+            acceptable TLS version. If either argument is ``None``, this will
+            leave that bound unchanged.
+            """
+
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket, server_side=False: bool,
+                        auto_handshake=True: bool,
+                        server_hostname=None: Optional[str]) -> TLSWrappedSocket:
```

Comment:
The real question is whether we need to doff our cap at the `ssl` module at all. If it didn't exist, would you care that `suppress_ragged_eofs` wasn't on this function?

As to `session`, right now I don't think that users should be in charge of session reuse. I think that error exists in the `ssl` module API due to historical quirks, and we shouldn't replicate it here.